### PR TITLE
Spending report

### LINKED ISSUE #35: Budget spending history
- research 

- simple example 

- [x] per account spending 
- [x] per tag spending (waiting for the tags become active)

### PR DESCRIPTION
**Pull request by guser21**
closes #35  

Fetches the blended costs for tags and users from AWS

### CODE REVIEW

**Comment by mbalc** [comment]
cause right now I can review how the code looks, but I'm totally out of context of the problem you're solving here, and I don't have a certain way to recap on that - that's why it's important to either name the PR and or describe it well
plus, the annotation `closes #35` automatizes issue management

**Comment by Ogochi** [comment]
@guser21 please update PR description: what task do you close and what actually have u done

**Comment by Ogochi** [comment]
Please make this PR close Issue not this PR :)

File: server/src/main/java/com/omigost/server/aws/CostService.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    //tags should be activated from amazon account
+    public List<AWSDailySpendingDTO> getSpendingForTags(DateInterval interval, Map<String, List<String>> tagValues) {
+        Expression accumulativeExpression = new Expression();
+
+        tagValues.entrySet().stream()
+                .findFirst()
+                .map(entry-> new TagValues().withKey(entry.getKey()).withValues(entry.getValue()))
+                .ifPresent(accumulativeExpression::setTags);
+
+        tagValues.entrySet().stream()
```
Comment (mbalc) [inline server/src/main/java/com/omigost/server/aws/CostService.java]: that looks ugly - I see that it works, don't see any simple way to fix this, but wanted to express my feelings about this need to handle the first tagValue here separately

Comment (mbalc) [reply server/src/main/java/com/omigost/server/aws/CostService.java]: would be nice if we could have empty expression and then call withAnd on it with the collection like this as argument, would be great if you tested this (there is support for collections in withAnd argument, but I'm not sure how the base Expression would behave - so sth along those lines
```
        new Expression().withAnd(
            tagValues.entrySet().stream()
                    .map(entry-> new TagValues().withKey(entry.getKey()).withValues(entry.getValue()))
                    .map(tag ->new Expression().withTags(tag) )
                    .toList()
                );
```

File: server/src/main/java/com/omigost/server/aws/OrganizationService.java [hunk 2]
```diff
@@ -34,12 +35,13 @@ private void initializeOrganizationClient() {
                 .withCredentials(credentialsProvider)
                 .withRegion(region)
                 .build();
+
     }
 
-    public void fetchAccounts() {
+    public List<Account> fetchAccounts() {
```
Comment (mbalc) [inline server/src/main/java/com/omigost/server/aws/OrganizationService.java]: +1

File: server/src/main/java/com/omigost/server/aws/CostService.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+import javax.annotation.PostConstruct;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Service
+public class CostService {
+    @Value("${aws.region}")
+    private String region;
```
[the remark this answers is not in the document]
Comment (guser21) [reply server/src/main/java/com/omigost/server/aws/CostService.java]: no Spring initializes them after the class is created

File: server/src/main/java/com/omigost/server/aws/CostService.java (hunk 1 above)
Comment (guser21) [reply server/src/main/java/com/omigost/server/aws/CostService.java]: None of those work. AWS throws exceptions on those requests. Refactored a little bit.

File: server/src/main/java/com/omigost/server/rest/AWSTestController.java [hunk 4]
```diff
@@ -17,14 +23,25 @@
...
-        return organization.getAccounts();
+        return organization.fetchAccounts();
     }
 
     @GetMapping("/budgets")
     public List<Budget> budgets() {
         return budgets.getBudgets();
     }
+
+    @PostMapping("/spending/account")
```
[the remark this answers is not in the document]
Comment (mbalc) [reply server/src/main/java/com/omigost/server/rest/AWSTestController.java]: what about this?

[the remark this answers is not in the document]
Comment (guser21) [reply server/src/main/java/com/omigost/server/rest/AWSTestController.java]: I should send a Post

[the remark this answers is not in the document]
Comment (mbalc) [reply server/src/main/java/com/omigost/server/rest/AWSTestController.java]: ok, it's not easy to send this kind of payload via GET

**Pull request merged**